src/sagas/deposit.js

```diff
@@ -70,6 +70,14 @@ function* restoreState(nextStepMap, stateKey) {
             // Explicitly fall through.
 
         case tbtc.Deposit.State.AWAITING_WITHDRAWAL_SIGNATURE:
+            yield put({
+                type: DEPOSIT_STATE_RESTORED,
+                payload: {
+                    depositAddress,
+                }
+            })
```

Comment:
This shouldn't be necessary, since the switch fallthrough should run through `DEPOSIT_STATE_RESTORED` below, I believe.

Reply:
I added this here because when transitioning from `/redemption` to `/signing` it ended up in a state in which the "Loading..." message was shown instead of the `<Signing />` component after adding the `resumeRedemption` to this case. It happened multiple times but I'll verify again.

Reply:
Without adding the `DEPOSIT_STATE_RESTORED` here, the UI gets stuck with the "Loading..." message in the case because while the switch does fall through to the second `DEPOSIT_STATE_RESTORED` below, it runs the `yield* finalCalls()` first, and in since we set it to `resumeRedemption` as the `finalCalls`, it has to wait for all the promises to resolve in that saga there it can move on to flagging the state as restored.

Reply:
However, there is some duped stuff here with the `AWAITING_WITHDRAWAL_PROOF` case so I'll clean that up.

Reply:
Maybe since the only time we seem to set `finalCalls` is when we `resumeRedemption`, I think I'll move that after `DEPOSIT_RESTORED` below and get rid of the one here??

Reply:
(Not doing the above, because it causes the app to cycle backwards to the wrong url after runRedemption completes)

Reply:
I'm guessing https://github.com/keep-network/tbtc-dapp/pull/224/commits/fc380ae01f448dd5efee1143c72a7b0693a7cd8f was the adjusted fix? I think now that you've gone deep on this whole thing, the next question is “how do we make this state restoration stuff not suck”. The current version is pretty slapped together, and I don't think it's really optimal.

> since we set it to resumeRedemption as the finalCalls, it has to wait for all the promises to resolve in that saga there it can move on to flagging the state as restored.

This seems like something we could adjust by splitting into two sagas, perhaps?

----

For the record, here's my mental model of what I really want the loading stuff to look like:

- Read an initial state from the URL.
- Resolve the real state on-chain. This isn't a UI-updating action, it's just about catching the UI state up.
- Once our current state matches the on-chain state, display the UI in that state.

Reply:
Yea - fc380ae was to address this. I agree, we should re-jigger how state restoration is handled generally, and maybe adding another saga for what to do after state is restored is the better answer than sticking it in the view. Thanks for peeping my commits, I was kind of debating back and forth whether I liked what I did, but I think maybe you're right we should just handle this an upcoming PR, like you mentioned in the other comment.

I also agree with your "mental loading model". When working on all this I noticed the issue you mentioned a while back about how if you reload, the UI ends up cycling through all the previous steps before we catch up to the current state.